pkg/collector/corechecks/snmp/snmp.go

```diff
@@ -26,27 +28,68 @@ type Check struct {
 	core.CheckBase
 	config         *checkconfig.CheckConfig
 	singleDeviceCk *devicecheck.DeviceCheck
+	discovery      discovery.Discovery
 }
 
 // Run executes the check
 func (c *Check) Run() error {
-
+	var checkErr error
 	sender, err := aggregator.GetSender(c.ID())
 	if err != nil {
 		return err
 	}
 
-	c.singleDeviceCk.SetSender(report.NewMetricSender(sender))
+	if c.config.IsDiscovery() {
+		var discoveredDevices []*devicecheck.DeviceCheck
+		discoveredDevices = c.discovery.GetDiscoveredDeviceConfigs()
+
+		jobs := make(chan *devicecheck.DeviceCheck, len(discoveredDevices))
+
+		var wg sync.WaitGroup
+
+		for w := 1; w <= c.config.Workers; w++ {
+			wg.Add(1)
+			go c.runCheckDeviceWorker(w, &wg, jobs)
+		}
+
+		for i := range discoveredDevices {
+			deviceCk := discoveredDevices[i]
+			deviceCk.SetSender(report.NewMetricSender(sender))
+			jobs <- deviceCk
+		}
+		close(jobs)
+		wg.Wait() // wait for all workers to finish
```

Comment:
isn't there a way to stop the check earlier?

Reply:
Sorry, I'm not sure to understand. Why stop the check earlier ? In which case ?

Reply:
I wondered if we needed to have early stop, as we do for discovery. Here if you have a `\16` subnet you will have to wait for all 65k devices to end their check before being able to stop. If the caller of `check.Cancel` has a timeout, this might lead to these runners to keep on running or to be halted if the process is killed on timeout 🤷‍♀️

Reply:
Indeed, for high number of devices it can help to stop earlier, but I think that can be an improvement for future if that's an issue.